Implement TryFrom

tcharding opened the pull request:
Audit the whole codebase checking for any method that is of the form `from_foo` where foo is not an interesting identifier (like 'consensus' and 'standard'). Implement `TryFrom` for any such methods, deprecating the original.

Done as separate patches so any can be easily dropped if not liked.

apoelstra reviewed (dismissed):
ACK c898301d0eec9805b8822579e0d5381b8b2a5348

sanket1729 commented on src/util/sighash.rs:
```diff
@@ -334,18 +336,29 @@ impl SchnorrSighashType {
+            0x81 => AllPlusAnyoneCanPay,
+            0x82 => NonePlusAnyoneCanPay,
+            0x83 => SinglePlusAnyoneCanPay,
+            0xFF => Reserved,
```
Opened a new issue about removing the variant here

sanket1729 reviewed (approved):
LGTM, but needs rebase

tcharding commented:
Changes in force push:
- rebased .. twice
- Used `to_u8` instead of deprecated `into_u8`

apoelstra reviewed (approved):
ACK 602c17acd2a896bee07f5ba1f39aed62b3d8c7ea

Kixunil commented on src/util/address.rs:
```diff
@@ -350,6 +327,102 @@ impl WitnessVersion {
+    fn try_from(opcode: opcodes::All) -> Result<Self, Self::Error> {
+        match opcode.to_u8() {
+            0 => Ok(WitnessVersion::V0),
+            version if version >= opcodes::all::OP_PUSHNUM_1.to_u8() && version <= opcodes::all::OP_PUSHNUM_16.to_u8() =>
```
Doesn't clippy complain about this? [I don't like](https://internals.rust-lang.org/t/clippy-suggestion-range-contains-may-cause-yoda-condition-is-my-fix-a-good-idea/16831) the lint much anyway, just surprised.

tcharding replied:
Super bizarre, it seems that to trigger the 'manual range' lint there has to be an `=` sign. This does _not_ trigger the lint
```rust
        let version = 0u8;
        if version > 1 && version < 10 {
            in_range = true;
        }
```
But this does (added `=`):
```rust
        let version = 0u8;
        if version >= 1 && version < 10 {
            in_range = true;
        }
```

Kixunil commented on src/util/taproot.rs:
```diff
@@ -718,6 +716,23 @@ impl TaprootMerkleBranch {
+            Ok(TaprootMerkleBranch(v))
+        }
+    }
+}
```
It'd be useful to add `TryFrom<&[sha256::Hash]>` since this is where `TryFrom` shines - efficiently handling both owned and non-owned conversions.

tcharding replied:
I didn't have this in mind when using `TryFrom`, thanks for the lesson. Added as suggested.

Kixunil commented on src/util/sighash.rs:
```diff
@@ -329,17 +331,28 @@ impl SchnorrSighashType {
+            0x82 => NonePlusAnyoneCanPay,
+            0x83 => SinglePlusAnyoneCanPay,
+            x => return Err(Error::InvalidSighashType(x as u32)),
+        })
```
Shouldn't this be `from_consensus_u8`?

tcharding replied:
I do not know how Bitcoin Core handles a a schnorr sighash type that is _not_  0-3 or 81-83. Looking solely at our code, for `EcdsaSighashType` we have `from_standard` and `from_consensus`, the first returns an error the later is infallible. If we are to use consensus/standard as part of the identifier for `SchnorrSighashType` should it not be the same, or does schnorr sighash not have this concept of standard vs consensus?

sanket1729 replied:
Any sighash type that is not 0-3 or 81-83 is rejected by consensus rules.

tcharding replied:
Ok, two opinions - check!

Will re-name to `from_consensus_u8`, thanks.

Kixunil reviewed (commented):
Would've acked but I'm not sure about sighash.

tcharding commented:
Changes in force push:
- Added the additional `TryFrom<&[sha256::Hash]>` version as suggested
- rebased

Remaining open question: what to do about sighash? I do think this PR is a move in the correct direction, possibly we could tackle the naming of `try_from` vs `from_consensus` at a later date? @Kixunil do you think we can find consensus quickly/easily or is it a thorny one?

Kixunil commented on src/util/taproot.rs:
```diff
@@ -718,6 +716,37 @@ impl TaprootMerkleBranch {
+    ///
+    /// If inner proof length is more than [`TAPROOT_CONTROL_MAX_NODE_COUNT`] (128).
+    fn try_from(v: &[sha256::Hash]) -> Result<Self, Self::Error> {
+        let v = v.to_vec();
```
This wasn't my point, the caller can already do this. :) My point was to check the length here so that allocation is avoided in the error case. If you want to avoid duplicating checking logic I have a trick for that:

```rust
// non-public
fn from_collection<T: AsRef<[sha256::Hash]> + Into<Vec<sha256::Hash>>>(collection: T) -> Result<Self, TaprootError> {
    if collection.as_ref().len() > TAPROOT_CONTROL_MAX_NODE_COUNT {
        Err(TaprootError::InvalidMerkleTreeDepth(collection.as_ref().len()))
    } else {
        Ok(TaprootMerkleBranch(collection.into()))
    }
}

// just directly call `Self::from_collection(v)` in all `TryFrom` impls
```

You could also use macros for `TryFrom` and also add `Box<[sha256::Hash]>` as well.

tcharding replied:
Nice trick :) Implemented as suggested, thanks.

tcharding commented:
Changes in force push: Add macro as suggested and squash into the final commit.

tcharding commented:
Weird, the CI run is not showing?

apoelstra reviewed (dismissed):
ACK 6ef1eccf21667287bd2e2c658139a52c84ca4f8b

tcharding commented:
Changes in force push:
- rebased on master
- Removed `TryFrom<u8> for SchnorrSighashType` patch, replaced with "rename `from_u8` -> `from_consensus_u8`" patch
- Re-ordered patches so the `TryFrom` ones were all together and the re-name was last.

apoelstra commented:
Intermediate commits seem not to compile now.

tcharding commented:
Force push is rebase to pick up clippy fix from #1063, no changes.

apoelstra reviewed (approved):
ACK b29ff9b71524882ea29061a1831a0b12d8788c62

(Having local compilation errors but just due to https://github.com/rust-bitcoin/rust-secp256k1/pull/466 )

Kixunil reviewed (approved):
ACK b29ff9b71524882ea29061a1831a0b12d8788c62
